src/nvim/ex_docmd.c

```diff
@@ -9454,22 +9454,35 @@ static void ex_folddo(exarg_T *eap)
 
 static void ex_terminal(exarg_T *eap)
 {
-  // We will call termopen() with ['shell'] if not given a {cmd}.
-  char *name = (char *)p_sh;
-  char *lquote = "['";
-  char *rquote = "']";
+  // We will call termopen() with shell as an arglist if not given a {cmd}.
+  char ex_cmd[1024];
+
   if (*eap->arg != NUL) {
-    name = (char *)vim_strsave_escaped(eap->arg, (char_u *)"\"\\");
-    lquote = rquote = "\"";
-  }
+    char *name = (char *)vim_strsave_escaped(eap->arg, (char_u *)"\"\\");
+    snprintf(ex_cmd, sizeof(ex_cmd),
+             ":enew%s | call termopen(\"%s\") | startinsert",
+             eap->forceit == true ? "!" : "", name);
+    xfree(name);
+  } else {
+    char **argv = build_argv((char *)p_sh);
+    char **p = argv;
 
-  char ex_cmd[512];
-  snprintf(ex_cmd, sizeof(ex_cmd),
-           ":enew%s | call termopen(%s%s%s) | startinsert",
-           eap->forceit==TRUE ? "!" : "", lquote, name, rquote);
-  do_cmdline_cmd(ex_cmd);
+    char shell_argv[512];
+    if (*p != NULL) {
+      snprintf(shell_argv, sizeof(shell_argv), "\"%s\"", *p);
+      p++;
+    }
 
-  if (name != (char *)p_sh) {
-    xfree(name);
+    while (*p != NULL) {
+      snprintf(shell_argv, sizeof(shell_argv), "%s, \"%s\"", shell_argv, *p);
```

Comment:
This seems like a weird way to concat a string. `vim_strcat` and `concat_str` are alternatives.